storage/imxrt-flash/flashdrv.c

```diff
@@ -119,7 +119,68 @@ static ssize_t bufferSync(flash_context_t *ctx, uint32_t dstAddr, int isLast)
 }
 
 
-ssize_t flash_bufferedPagesWrite(flash_context_t *ctx, uint32_t dstAddr, const void *srcPtr, size_t size)
+ssize_t flash_bufferedRead(flash_context_t *ctx, uint32_t srcAddr, void *dstPtr, size_t size)
+{
+	ssize_t res;
+
+	if ((ctx->prevAddr == (uint32_t)-1) || (ctx->isDirty == 0)) {
+		return flash_directRead(ctx, srcAddr, dstPtr, size);
+	}
+
+	uint32_t bufSize = ctx->properties.sector_size;
+	uint32_t bufAddr = ctx->prevAddr & ~(bufSize - 1);
+	uint32_t bufOffs = srcAddr & (bufSize - 1);
+
+	/* Source area does not overlap the buffer window */
+	if ((srcAddr >= bufAddr + bufSize) || (bufAddr >= srcAddr + size)) {
+		flash_directRead(ctx, srcAddr, dstPtr, size);
+	}
+	/* Source area overlaps the buffer window on the bottom */
+	else if ((srcAddr < bufAddr) && (srcAddr + size > bufAddr)) {
+		uint32_t len = bufAddr - srcAddr;
+		res = flash_directRead(ctx, srcAddr, dstPtr, len);
+		if (res < 0) {
+			return res;
+		}
+		memcpy(dstPtr, ctx->buff, size - len);
+	}
+	/* Source area contains within the buffer window (inner) */
+	else if ((srcAddr >= bufAddr) && (srcAddr + size <= bufAddr + bufSize)) {
+		memcpy(dstPtr, ctx->buff + bufOffs, size);
+	}
+	/* Source area overlaps the buffer window on the top */
+	else if ((srcAddr < bufAddr + bufSize) && (srcAddr + size > bufAddr + bufSize)) {
+		uint32_t bufEndAddr = bufAddr + bufSize;
+		uint32_t len = bufEndAddr - srcAddr;
+		memcpy(dstPtr, ctx->buff + bufOffs, len);
+		res = flash_directRead(ctx, bufEndAddr, dstPtr, size - len);
+		if (res < 0) {
+			return res;
+		}
+	}
+	/* Buffer window contains within the source area (outer) */
+	else if ((bufAddr >= srcAddr) && (bufAddr + bufSize <= srcAddr + size)) {
+		uint32_t len = bufAddr - srcAddr;
+		uint32_t outerSize = size - bufSize;
+
+		res = flash_directRead(ctx, srcAddr, dstPtr, len);
+		if (res < 0) {
+			return res;
+		}
+		memcpy(dstPtr, ctx->buff, bufSize);
+		res = flash_directRead(ctx, bufAddr + bufSize, dstPtr, outerSize - len);
+		if (res < 0) {
+			return res;
+		}
+	}
+	else {
+	}
```

Comment:
It it's really impossible to don't fall into any of `if`s above, then maybe last case should be `else`? On the other hand, if it is possible indeed, then maybe we should e.g. return error here?

Reply:
It's because of MISRA rule 15.7 _"All if … else if constructs shall be terminated with an else statement"_ which is required, _"a final else statement is not required for a simple if statement"_, it is like requirement  with `default` case in `switch-case` construct.

It's unreachable I agree, compiler removes it.. of course I may add return error code with /* unreachable */ comment. Compliant example from MISRA:
```C
if (..) {
  /* ... */
} else if (..) {
  /* ... */
} else {
  ; /* No action required - `;` is optional */
}
```

Reply:
I'm aware why you added it, but maybe we don't need to check anything in the last `else if` and replace it with `else` then

Reply:
The function has changed shape (rewritten).